src/boot/efi/initrd.c

```diff
@@ -0,0 +1,135 @@
+        }
+
+        CopyMem(buffer, loader->address, loader->length);
+        *buffer_size = loader->length;
+        return EFI_SUCCESS;
+}
+
+EFI_STATUS initrd_register(
+                VOID *initrd_address,
+                UINTN initrd_length,
+                EFI_HANDLE *initrd_handle) {
+
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *dp = (EFI_DEVICE_PATH *) &efi_initrd_device_path;
+        EFI_HANDLE handle;
+        struct initrd_loader *loader;
+
+        assert(initrd_handle);
+
+        if (!initrd_address || initrd_length == 0)
+                return EFI_SUCCESS;
+
+        /* check if a InitrdMedia DevicePath is alreay registed.
+           LocateDevicePath checks for the "closest DevicePath" and returns its handle,
+           where as InstallMultipleProtocolInterfaces only maches identical DevicePaths.
+         */
+        err = uefi_call_wrapper(BS->LocateDevicePath, 3, &EfiLoadFile2Protocol , &dp, &handle);
+        if (err != EFI_NOT_FOUND) /* InitrdMedia is already registered */
+                return EFI_ALREADY_STARTED;
+
+        loader = AllocatePool(sizeof(struct initrd_loader));
+        if (!loader)
+                return EFI_LOAD_ERROR;
+
+        loader->load_file.LoadFile = initrd_load_file;
+        loader->address = initrd_address;
+        loader->length = initrd_length;
+
+        /* create a new handle and register the LoadFile2 protocol with the InitrdMediaPath on it */
+        err = uefi_call_wrapper(BS->InstallMultipleProtocolInterfaces, 8,
+                                initrd_handle,
+                                &DevicePathProtocol, &efi_initrd_device_path,
+                                &EfiLoadFile2Protocol, loader,
+                                NULL);
+        if (EFI_ERROR(err)) {
+                *initrd_handle = NULL;
+                FreePool(loader);
+        }
+
+        return err;
+}
+
+EFI_STATUS initrd_deregister(EFI_HANDLE initrd_handle) {
+        EFI_STATUS err;
+        struct initrd_loader *loader;
+
+        if (initrd_handle) {
+                /* get the LoadFile2 protocol that we allocated earlier */
+                err = uefi_call_wrapper(BS->OpenProtocol, 3,
+                                        initrd_handle, &EfiLoadFile2Protocol, (VOID **) &loader,
+                                        NULL, NULL, EFI_OPEN_PROTOCOL_GET_PROTOCOL);
+                if (EFI_ERROR(err))
+                        return EFI_SUCCESS;
+
+                /* uninstall all protocols thus destroying the handle */
+                err = uefi_call_wrapper(BS->UninstallMultipleProtocolInterfaces, 6,
+                                        initrd_handle,
+                                        &DevicePathProtocol, &efi_initrd_device_path,
+                                        &EfiLoadFile2Protocol, loader,
+                                        NULL);
+
+                if (!EFI_ERROR(err))
+                        FreePool(loader);
```

Comment:
doesn't matter, but i'd probably inver the logic so that the exception is handled in the if block and the clean case follows the rest of the function. i.e. just do `if (EFI_ERROR(err)) return ern;` here, and then do `FreePool()` as last step before `return EFI_SUCCESS`